hazelcast/src/test/java/com/hazelcast/spi/impl/executionservice/impl/BasicCompletableFutureTest.java

```diff
@@ -0,0 +1,296 @@
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.rules.ExpectedException;
+
+import java.util.concurrent.Callable;
+import java.util.concurrent.Future;
+import java.util.concurrent.FutureTask;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+import static org.mockito.Matchers.any;
+import static org.mockito.Matchers.anyString;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyZeroInteractions;
+import static org.mockito.Mockito.when;
+
+@Category(QuickTest.class)
+public class BasicCompletableFutureTest {
+
+    private static final String DELEGATE_RESULT = "DELEGATE_RESULT";
+    private static final String OUTER_RESULT = "OUTER_RESULT";
+
+    @Rule
+    public ExpectedException expected = ExpectedException.none();
+
+    private FutureTask<String> delegateFuture;
+    private boolean delegateThrowException;
+    private BasicCompletableFuture<String> outerFuture;
+
+    @Before
+    public void given() {
+        delegateThrowException = false;
+        delegateFuture = future(DELEGATE_RESULT);
+        outerFuture = basicCompletableFuture(delegateFuture);
+    }
+
+    @Test
+    public void cancel_delegate_bothCancelled() {
+        delegateFuture.cancel(false);
+
+        assertTrue(delegateFuture.isCancelled());
+        assertTrue(outerFuture.isCancelled());
+    }
+
+    @Test
+    public void cancel_outer_bothCancelled() {
+        outerFuture.cancel(false);
+
+        assertTrue(delegateFuture.isCancelled());
+        assertTrue(outerFuture.isCancelled());
+    }
+
+    @Test
+    public void completeOuter_outerDoneDelegateNot_delegateAskedFirst() {
+        outerFuture.setResult(OUTER_RESULT);
+
+        assertFalse(delegateFuture.isDone());
+        assertTrue(outerFuture.isDone());
+    }
+
+    @Test
+    public void completeOuter_outerDoneDelegateNot_outerAskedFirst() {
+        outerFuture.setResult(OUTER_RESULT);
+
+        assertTrue(outerFuture.isDone());
+        assertFalse(delegateFuture.isDone());
+    }
+
+    @Test
+    public void completeDelegate_bothDone_delegateAskedFirst() {
+        delegateFuture.run();
+
+        assertTrue(delegateFuture.isDone());
+        assertTrue(outerFuture.isDone());
+    }
+
+    @Test
+    public void completeDelegate_bothDone_outerAskedFirst() {
+        delegateFuture.run();
+
+        assertTrue(outerFuture.isDone());
+        assertTrue(delegateFuture.isDone());
+    }
+
+    @Test
+    // PROBLEM NR.1
+    // If you complete the BasicCompletableFuture through setResult it will return isDone() == true,
```

Comment:
This is the current idea: AbstractCompletableFuture should not expose setResult, it should be a protected method. Note that ICompletableFuture actually defines the API of a _listenable_,  _noncompletable_ future. This will now cause major confusion as we also plan a subclass of ACF which exposes setResult -- what should we name it? ThisTimeForRealCompletableFuture?